tracecat/llm/ollama.py

```diff
@@ -0,0 +1,129 @@
+"""Ollama LLM client.
+
+Docs: https://github.com/ollama/ollama/blob/main/docs/api.md
+"""
+
+from enum import StrEnum
+from typing import Any
+
+import ollama
+from ollama import ChatResponse
+from pydantic import BaseModel
+
+from tracecat import config
+from tracecat.logger import logger
+
+
+# NOTE: We specify the params size tag otherwise
+# it defaults to the ambigious "latest" tag
+# https://ollama.com/search
+class OllamaModel(StrEnum):
+    # Smol models (<15GB)
+    LLAMA32 = "llama3.2:3b"
+    LLAMA32_1B = "llama3.2:1b"
+    MISTRAL_SMALL = "mistral-small:24b"
+    # Large models
+    LLAMA33 = "llama3.3:70b"
+    MISTRAL_LARGE = "mistral-large:123b"
+    MIXTRAL = "mixtral:8x7b"
+    MIXTRAL_22B = "mixtral:8x22b"
+
+
+DEFAULT_OLLAMA_MODEL = OllamaModel.LLAMA32_1B
+
+
+def _get_ollama_client(api_url: str | None = None) -> ollama.AsyncClient:
+    return ollama.AsyncClient(host=api_url or config.OLLAMA__API_URL)
+
+
+async def preload_ollama_models(models: list[str]) -> list[dict[str, Any]]:
+    client = _get_ollama_client()
+    responses = []
+    # Download iteratively to avoid overwhelming
+    # the Ollama client and server
+    for model in models:
+        try:
+            await client.pull(model)
+        except Exception as e:
+            logger.warning(
+                "Failed to pull model",
+                model=model,
+                error=e,
+            )
+    return responses
```

Comment:
interesting, isn't this operation IO bound? does it actually overwhelm ollama?

Reply:
Some users in network constrained environments faced some issues pulling multiple models at the same time. Yes it's IO bound but Ollama loads models in chunks and sometimes weird loss chunks and errors occur?

Found pulling one by one fixed it.